### ISSUE TITLE
About fragment inference

### ISSUE THREAD

**Issue by Terminal-K** [Pointcept/Pointcept#164]
Hi, @Gofinge 
I was doing fragment inference and found that it was very slow. When I checked this part of the code, I found "fragment_batch_size".
I increased it from 1 to 8 and found that it would improve the overall speed, but I don't know if it would have any other effect.
https://github.com/Pointcept/Pointcept/blob/553bf5a1520d7e9326aec4c602f42bf50b76b4c8/pointcept/engines/test.py#L176-L180

BTW, can you tell me roughly how different the results of fragment inference will be from the results of validation during the training? 
I wonder if my differences are similar to yours.

Thanks.

**Comment by Gofinge** [comment]
Fragment inference only supports bs1, yet there exists an acceleration strategy, please refer here (https://github.com/Pointcept/Pointcept/issues/108#issuecomment-1893709522).

**Comment by Terminal-K** [comment]
Thanks for your reply! 
I had tried this approach you said before, but the testing was still slow (~ 30 hours, with 5e5 points per scan), so I tried to find improvements from fragment inference. When I increase the size of fragment_batch_size, the program seems to work, but I don't know if the final performance will be affected compared to fragment_batch_size=1.

**Comment by Terminal-K** [comment]
It looks like I could try testing without sampling

**Comment by Terminal-K** [comment]
I was wrong :( 
The metrics obtained without downsampling were very different from the validation results of the training process, which I think is probably due to grid_size and serialization. However, I have now increased the fragment_batch_size of fragment inference, and now the metrics are normal so far.

**Comment by Gofinge** [comment]
> I was wrong :( The metrics obtained without downsampling were very different from the validation results of the training process, which I think is probably due to grid_size and serialization. However, I have now increased the fragment_batch_size of fragment inference, and now the metrics are normal so far.

Currently, there is no existing backbone robust to different sampling densities, fix the grid size as your training setting is necessary.

> I had tried this approach you said before, but the testing was still slow (~ 30 hours, 5e5 points per scan, 4000 scans in total), so I tried to find improvements from fragment inference. When I increase the size of fragment_batch_size, the program seems to work, but I don't know if the final performance will be affected compared to fragment_batch_size=1.

Could you provide your detailed config for further diagnosis?

**Comment by Terminal-K** [comment]
Sorry for the late reply. Here's my original data config:
![datacfg_train2](https://github.com/Pointcept/Pointcept/assets/59523614/c4991e11-006e-4e2c-8a74-62fd40e92ccb)
![datacfg_val2](https://github.com/Pointcept/Pointcept/assets/59523614/009d09be-a231-4157-9887-3b8d7ca25d2d)
![datacfg_test2](https://github.com/Pointcept/Pointcept/assets/59523614/03cc7b8e-e8ef-4c9b-8bf0-96afb68e5b55)
However, I was concerned that using the first GridSample would affect the final performance, and in my most recent experiment I didn't use it and increased "fragment_batch_size" instead. However, the results of this experiment fragment inference(miou=88.85%) are 5 points lower on the validation set than the miou of the validation performed during training(miou=92.35%). I wonder if this level of performance loss is normal.

Looking forward to your reply and wishing you all the best :)

**Comment by Gofinge** [comment]
Hi, I briefly went through the configs you wrote, so here is a suggestion:
1. I base grid size of 0.05 is suggested for outdoor scenes, yet 0.1 should also be okay.
2. During the test process, the pre-voxelization should set a grid size half of the base grid size. In your case, please use 0.05 instead of 0.025 for a base grid size of 0.1.
3. I still have not verified the effectiveness of fragment_batch_size.
4. Usually no need to worry the pre-voxelization will have a negative influence on the final results. In my ablation, it works well.

**Comment by Terminal-K** [comment]
Thank you very much for your reply!
I will try to change the grid size of pre-voxelization to 0.05 and set fragment_batch_size to 1 for testing again. Subsequent results will be reported here.

**Comment by Terminal-K** [comment]
Hi, @Gofinge 
I used the pre-voxelized method you mentioned for testing. The miou of fragment inference on the validation differs from the validation miou of the training process (92.35%->90.11%) by about 2 points, with an iou drop of 2.7 points (77.87%->75.14%) for the most critical class, which seems more acceptable. The reason for the 5-point loss mentioned earlier was a small bug in my configuration somewhere. However, the results on iou are not too different from what I did earlier when I added fragment_batch_size. However, the difference is that fragment inference in this way does spend much less time.
I am very sorry to disturb you due to my mistake, and thank you very much for your earnest reply, which really helped me a lot. Wish you all the best :)

**Comment by Gofinge** [comment]
Great, it seems that fragment_batch_size does work. The evaluation results in not accurate as data during evaluation is down-sampled. Also, we have an interpolation strategy for the evaluation process to acquire precise performance. You can refer val config for S3DIS (https://github.com/Pointcept/Pointcept/blob/main/configs/s3dis/semseg-pt-v3m1-0-rpe.py#L139). (Check "Copy" and "Collect")
